crates/oxc_minifier/src/peephole/minimize_statements.rs

```diff
@@ -443,6 +446,254 @@ impl<'a> PeepholeOptimizations {
+            && id_kind.rest.is_none()
+            && !id_kind.elements.is_empty()
+        {
+            return false;
+        }
+
+        // check for `[a = b] = [c]`, this could be potentially optimized if we check if `c` is undefined
+        if init_expr.elements.len() == 1
+            && id_kind
+                .elements
+                .first()
+                .is_some_and(|e| e.as_ref().is_none_or(BindingPattern::is_assignment_pattern))
+            && first.is_none_or(|expr| expr.value_type(ctx).is_undetermined())
+        {
+            return false;
+        }
+
+        true
+    }
+
+    fn simplify_array_destruction_assignment(
+        decl: &mut VariableDeclarator<'a>,
+        result: &mut Vec<'a, VariableDeclarator<'a>>,
+        ctx: &mut Ctx<'a, '_>,
+    ) -> bool {
+        let BindingPattern::ArrayPattern(id_pattern) = &mut decl.id else {
+            return false;
+        };
+        let Some(Expression::ArrayExpression(init_expr)) = &mut decl.init else {
+            return false;
+        };
+
+        // limit iteration of id (left) to last spread of init (right)
+        // [a, b, c] = [b, ...spread]
+        let index = if let Some(spread_index) =
+            init_expr.elements.iter().position(ArrayExpressionElement::is_spread)
+        {
+            min(id_pattern.elements.len(), spread_index)
+        } else {
+            id_pattern.elements.len()
+        };
+
+        if decl.kind.is_var() && index > 1 {
+            for init_val in &mut init_expr.elements {
+                if init_val.value_type(ctx).is_undetermined() {
+                    // create intermediate declarations to avoid issues with hoisting
+                    // `var [a] = [b]` => `var _a = b, a = _a`
+                    let ident =
+                        ctx.generate_uid_in_current_scope("_", SymbolFlags::FunctionScopedVariable);
+                    result.push(ctx.ast.variable_declarator(
+                        init_val.span(),
+                        VariableDeclarationKind::Let,
+                        ident.create_binding_pattern(ctx),
+                        NONE,
+                        Some(match init_val {
+                            ArrayExpressionElement::Elision(_) => ctx.ast.void_0(SPAN),
+                            _ => init_val.take_in(ctx.ast).into_expression(),
+                        }),
+                        decl.definite,
+                    ));
+                    *init_val = ArrayExpressionElement::from(ident.create_read_expression(ctx));
+                }
+            }
+        }
+
+        let mut init_iter = init_expr.elements.drain(..);
+
+        for id_item in id_pattern.elements.drain(0..index) {
+            let init_val = match init_iter.next() {
+                None | Some(ArrayExpressionElement::Elision(_)) => {
+                    ArrayExpressionElement::from(ctx.ast.void_0(SPAN))
+                }
+                Some(init_val) => init_val,
```

Comment:
```suggestion
                None | Some(ArrayExpressionElement::Elision(_)) => {
                    ctx.ast.void_0(SPAN)
                }
                Some(ArrayExpressionElement::SpreadElement(_)) => {
                    unreachable!("spread element does not exist until `index`")
                }
                Some(init_val) => init_val.to_expression(),
```
I think it'd be nice to explicitly handle SpreadElement here. I guess this would remove the need of `impl<'a> DetermineValueType<'a> for ArrayExpressionElement<'a>`.